equal.go

```diff
@@ -317,16 +317,22 @@ func (t *T) deepEqual(
 		// and assert to a string will panic.
 		hstr := have.String()
 		wstr := want.String()
-		if len(hstr) != len(wstr) {
+		if hstr == wstr {
+			// Cheap equality test passed, no need to continue.
+			break
+		}
+		hrunes := []rune(hstr)
+		wrunes := []rune(wstr)
+		if len(hrunes) != len(wrunes) {
 			return []string{
 				fmt.Sprintf("%s: len(have) %d != len(want) %d.",
-					desc, len(hstr), len(wstr)),
+					desc, len(hrunes), len(wrunes)),
 				fmt.Sprintf("  have: %#v", hstr),
 				fmt.Sprintf("  want: %#v", wstr),
 			}
 		}
-		for i := range hstr {
-			if hstr[i] != wstr[i] {
+		for i, r := range hrunes {
+			if r != wrunes[i] {
 				return []string{
 					fmt.Sprintf("%s: difference at index %d.", desc, i),
```

Comment:
nit: I think this indexing is now a little confusing; how is it indexed?  `s/index/rune/`?

Error message does not clearly guide the user to the correct point.